bleak/backends/bluezdbus/discovery.py

```diff
@@ -53,50 +53,178 @@ def _device_info(path, props):
+        return
+
+    uuids = props.get("UUIDs", [])
+    manufacturer_data = props.get("ManufacturerData", {})
+    return BLEDevice(
+            address,
+            name,
+            {"path": path, "props": props},
+            uuids=uuids,
+            manufacturer_data=manufacturer_data,
+        )
 
-    For possible values for `filter`, see the parameters to the
-    ``SetDiscoveryFilter`` method in the `BlueZ docs
-    <https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/adapter-api.txt?h=5.48&id=0d1e3b9c5754022c779da129025d493a198d49cf>`_
 
-    The ``Transport`` parameter is always set to ``le`` by default in Bleak.
+class AsyncDiscovery():
 
-    Args:
-        timeout (float): Duration to scan for.
-        loop (asyncio.AbstractEventLoop): Optional event loop to use.
+    def __init__(self, callback, loop=None, device="hci0"):
+        """State keeper to discover nearby Bluetooth Low Energy devices and get
+        a call for each discovered device in an asynchronous way.
 
-    Keyword Args:
-        device (str): Bluetooth device to use for discovery.
-        filters (dict): A dict of filters to be applied on discovery.
+        Args:
+            callback (Callable[[bleak.BLEDevice], None]): called for each discovered device.
+            loop (asyncio.AbstractEventLoop): Optional event loop to use.
+            device (str): Bluetooth device to use for discovery.
 
-    Returns:
-        List of tuples containing name, address and signal strength
-        of nearby devices.
+        """
+        self.callback = callback
+        self.device = device
+        self.loop = loop if loop else asyncio.get_event_loop()
+        self.rules = list()
+        self.cached_devices = {}
+        self.bus = None
+        self.devices = {}
+        self.adapter_path = ""
 
-    """
-    device = kwargs.get("device", "hci0")
-    loop = loop if loop else asyncio.get_event_loop()
-    cached_devices = {}
-    devices = {}
-    rules = list()
+    async def start_discovery(self, filters=None):
+        """Start discovering of nearby BLE devices.
 
-    # Discovery filters
-    filters = kwargs.get("filters", {})
-    filters["Transport"] = "le"
+        For possible values for `filters`, see the parameters to the
+        ``SetDiscoveryFilter`` method in the `BlueZ docs
+        <https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/adapter-api.txt?h=5.48&id=0d1e3b9c5754022c779da129025d493a198d49cf>`_
+
+        The ``Transport`` parameter is always set to ``le`` by default in Bleak.
+
+        The filters are applied and the callback registered with the object is
+        called every time a new device appears or the properties of an already
+        discovered device changes. This might happen frequently, since a change
+        in the RSSI value is considered a property change.
+
+        Args:
+            filters (dict): A dict of filters to be applied on discovery.
+
+        """
+        if self.rules:
```

Comment:
should these and other references to attributes of `self` not be guarded against concurrent invocations of `start_discovery()` and `stop_discovery()`? Or alternatively prevent such races?

Personally i would just move all state into a `Discovery` class and let every `start_discovery()` create and return it to avoid concurrency in the first place:
```
class AsyncDiscovery:
  ...

  def start_discovery(...):
     disco = Discovery(**self)
     _start_disco(disco)
     return disco

def _start_disco(disco):
  .. what start_discovery does

class Discovery:
  ...

  def stop_discovery(...):
    ...
```

Reply:
lgtm now :+1: